File: foundations-jdbc-kotlin/src/kotlin/dev/typr/foundationskt/Fragment.kt
@file:Suppress("unused")
package dev.typr.foundationskt

import java.util.Optional

class Fragment(val underlying: dev.typr.foundations.Fragment) {

    fun render(): String = underlying.render()

    override fun toString(): String {
        val ctx = SqlBuilder.currentContext()
        if (ctx != null) {
            return ctx.register(this)
        }
        return render()
    }

    fun append(other: Fragment): Fragment = Fragment(underlying.append(other.underlying))

    operator fun plus(other: Fragment): Fragment = append(other)

    fun <T> query(parser: ResultSetParser<T>): OperationRead.Query<T> =
        OperationRead.Query(dev.typr.foundations.OperationRead.Query(underlying, parser.underlying))

    fun <T : Any> queryExactlyOne(type: DbType<T>): OperationRead.Query<T> =
        query(RowCodec.of(type).exactlyOne())

    fun <T : Any> queryExactlyOne(codec: RowCodec<T>): OperationRead.Query<T> =
        query(codec.exactlyOne())

    fun <T : Any> queryAll(type: DbType<T>): OperationRead.Query<List<T>> =
        query(RowCodec.of(type).all())

    fun <T : Any> queryAll(codec: RowCodec<T>): OperationRead.Query<List<T>> =
        query(codec.all())

    fun <T : Any> queryMaxOne(type: DbType<T>): OperationRead.Query<T?> =
        query(RowCodec.of(type).maxOne())

    fun <T : Any> queryMaxOne(codec: RowCodec<T>): OperationRead.Query<T?> =
        query(codec.maxOne())

    fun update(): Operation.Update =
        Operation.Update(dev.typr.foundations.Operation.Update(underlying))

    fun execute(): Operation.Execute =
        Operation.Execute(dev.typr.foundations.Operation.Execute(underlying))

    fun <T> updateReturning(parser: ResultSetParser<T>): Operation.UpdateReturning<T> =
        Operation.UpdateReturning(dev.typr.foundations.Operation.UpdateReturning(underlying, parser.underlying))

    fun <T> updateReturningGeneratedKeys(columnNames: Array<String>, parser: ResultSetParser<T>): Operation.UpdateReturningGeneratedKeys<T> =
        Operation.UpdateReturningGeneratedKeys(underlying.updateReturningGeneratedKeys(columnNames, parser.underlying))

    fun <Row : Any> updateMany(parser: RowCodec<Row>, rows: Iterator<Row>): Operation.UpdateMany<Row> =
        Operation.UpdateMany(underlying.updateMany(parser.underlying, rows))

    fun <Row : Any> updateManyReturning(parser: RowCodec<Row>, rows: Iterator<Row>): Operation.UpdateManyReturning<Row> =
        Operation.UpdateManyReturning(underlying.updateManyReturning(parser.underlying, rows))

    fun <Row : Any> updateReturningEach(parser: RowCodec<Row>, rows: Iterator<Row>): Operation.UpdateReturningEach<Row> =
        Operation.UpdateReturningEach(underlying.updateReturningEach(parser.underlying, rows))

    fun <T : Any> streamingQuery(codec: RowCodec<T>, fetchSize: Int): OperationRead.Streaming<T> =
        OperationRead.Streaming(underlying.streamingQuery(codec.underlying, fetchSize))

    fun <T : Any> streamingQuery(type: DbType<T>, fetchSize: Int): OperationRead.Streaming<T> =
        OperationRead.Streaming(underlying.streamingQuery(type.underlying, fetchSize))

    fun append(s: String): Fragment = Fragment(underlying.append(s))

    fun <T> value(dbType: DbType<T>, value: T): Fragment = Fragment(underlying.value(dbType.underlying, value))

    fun appendAll(fragments: List<Fragment>, separator: Fragment): Fragment =
        Fragment(underlying.appendAll(fragments.map { it.underlying }, separator.underlying))

    fun <T : Any> valueNullable(dbType: DbType<T>, value: T?): Fragment =
        Fragment(underlying.value(dbType.underlying.opt(), Optional.ofNullable(value)))

    fun <Row : Any> row(parser: RowCodecNamed<Row>, row: Row, vararg except: String): Fragment =
        Fragment(underlying.row(parser.underlying, row, *except))

    fun <P0> param(dbType: DbType<P0>): ParamBuilders.ParamBuilder1<P0> =
        ParamBuilders.ParamBuilder1(underlying.param(dbType.underlying))

    // ── Conditional append DSL ──

    fun <T : Any> optionally(value: T?): OptionallyValue<T> = OptionallyValue(this, value)
    fun optionally(condition: Boolean): OptionallyFlag = OptionallyFlag(this, condition)

    class OptionallyValue<T : Any>(private val base: Fragment, private val value: T?) {
        fun append(sql: String, type: DbType<T>): Fragment {
            val javaOpt = base.underlying.optionally(java.util.Optional.ofNullable(value))
            return Fragment(javaOpt.append(sql, type.underlying))
        }

        fun append(sql: String, type: DbType<T>, whenAbsent: String): Fragment {
            val javaOpt = base.underlying.optionally(java.util.Optional.ofNullable(value))
            return Fragment(javaOpt.append(sql, type.underlying, whenAbsent))
        }
    }

    class OptionallyFlag(private val base: Fragment, private val condition: Boolean) {
        fun append(sql: String): Fragment =
            Fragment(base.underlying.optionally(condition).append(sql))

        fun append(fragment: Fragment): Fragment =
            Fragment(base.underlying.optionally(condition).append(fragment.underlying))

        fun append(whenTrue: String, whenFalse: String): Fragment =
            Fragment(base.underlying.optionally(condition).append(whenTrue, whenFalse))

        fun append(whenTrue: Fragment, whenFalse: Fragment): Fragment =
            Fragment(base.underlying.optionally(condition).append(whenTrue.underlying, whenFalse.underlying))
    }

    companion object {
        @JvmField
        val EMPTY: Fragment = Fragment(dev.typr.foundations.Fragment.EMPTY)

        @JvmStatic
        fun of(value: String): Fragment = Fragment(dev.typr.foundations.Fragment.of(value))

        @JvmStatic
        fun empty(): Fragment = EMPTY

        @JvmStatic
        fun quotedDouble(value: String): Fragment = Fragment(dev.typr.foundations.Fragment.quotedDouble(value))

        @JvmStatic
        fun quotedSingle(value: String): Fragment = Fragment(dev.typr.foundations.Fragment.quotedSingle(value))

        @JvmStatic
        fun <A> value(value: A, dbType: DbType<A>): Fragment =
            Fragment(dev.typr.foundations.Fragment.value(value, dbType.underlying))

        @JvmStatic
        fun <A> encode(dbType: DbType<A>, value: A): Fragment =
            Fragment(dev.typr.foundations.Fragment.encode(dbType.underlying, value))

        @JvmStatic
        fun and(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.and(fragments.map { it.underlying }))

        @JvmStatic
        fun and(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.and(fragments.map { it.underlying }))

        @JvmStatic
        fun or(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.or(fragments.map { it.underlying }))

        @JvmStatic
        fun or(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.or(fragments.map { it.underlying }))

        @JvmStatic
        fun whereAnd(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.whereAnd(fragments.map { it.underlying }))

        @JvmStatic
        fun whereAnd(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.whereAnd(fragments.map { it.underlying }))

        @JvmStatic
        fun whereOr(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.whereOr(fragments.map { it.underlying }))

        @JvmStatic
        fun whereOr(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.whereOr(fragments.map { it.underlying }))

        @JvmStatic
        fun set(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.set(fragments.map { it.underlying }))

        @JvmStatic
        fun set(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.set(fragments.map { it.underlying }))

        @JvmStatic
        fun parentheses(fragment: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.parentheses(fragment.underlying))

        @JvmStatic
        fun comma(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.comma(fragments.map { it.underlying }))

        @JvmStatic
        fun comma(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.comma(fragments.map { it.underlying }))

        @JvmStatic
        fun orderBy(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.orderBy(fragments.map { it.underlying }))

        @JvmStatic
        fun orderBy(fragments: List<Fragment>): Fragment =
            Fragment(dev.typr.foundations.Fragment.orderBy(fragments.map { it.underlying }))

        @JvmStatic
        fun join(fragments: List<Fragment>, separator: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.join(fragments.map { it.underlying }, separator.underlying))

        @JvmStatic
        fun concat(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.concat(*fragments.map { it.underlying }.toTypedArray()))

        @JvmStatic
        fun of(vararg fragments: Fragment): Fragment =
            Fragment(dev.typr.foundations.Fragment.of(*fragments.map { it.underlying }.toTypedArray()))

        /**
         * Emit `DROP TABLE IF EXISTS <table>`. Works on PostgreSQL, DuckDB, MariaDB, MySQL,
         * SQL Server (2016+) and Oracle (23c+). DB2 does not support the `IF EXISTS` clause —
         * wrap the plain `DROP TABLE <table>` in a try/catch for SQLSTATE 42704.
         */
        @JvmStatic
        fun dropTableIfExists(table: String): Operation.Execute =
            Operation.Execute(dev.typr.foundations.Fragment.dropTableIfExists(table))

        @JvmStatic
        fun <Row : Any> insertOne(
            table: String, codec: RowCodecNamed<Row>, row: Row, vararg except: String
        ): Operation.Update =
            Operation.Update(dev.typr.foundations.Fragment.insertOne(table, codec.underlying, row, *except))

        @JvmStatic
        fun <Row : Any> insertMany(
            table: String, codec: RowCodecNamed<Row>, rows: Iterator<Row>, vararg except: String
        ): Operation.BatchUpdate<Row> =
            Operation.BatchUpdate(dev.typr.foundations.Fragment.insertMany(table, codec.underlying, rows, *except))

        @JvmStatic
        fun <Row : Any> insertOneReturning(
            table: String, codec: RowCodecNamed<Row>, row: Row, vararg except: String
        ): OperationRead.Query<Row> =
            OperationRead.Query(dev.typr.foundations.Fragment.insertOneReturning(table, codec.underlying, row, *except))

        @JvmStatic
        fun <In : Any, Out : Any> insertOneReturning(
            table: String, writeCodec: RowCodecNamed<In>, row: In, readCodec: RowCodecNamed<Out>
        ): OperationRead.Query<Out> =
            OperationRead.Query(dev.typr.foundations.Fragment.insertOneReturning(
                table, writeCodec.underlying, row, readCodec.underlying))

        @JvmStatic
        fun <Row : Any, Out> insertOneGenerated(
            table: String, codec: RowCodecNamed<Row>, row: Row,
            generatedColumns: Array<String>, parser: ResultSetParser<Out>, vararg except: String
        ): Operation.UpdateReturningGeneratedKeys<Out> =
            Operation.UpdateReturningGeneratedKeys(dev.typr.foundations.Fragment.insertOneGenerated(
                table, codec.underlying, row, generatedColumns, parser.underlying, *except))

        @JvmStatic
        fun <Row : Any> upsertOne(
            table: String, codec: RowCodecNamed<Row>, row: Row, vararg conflictColumns: String
        ): Operation.Update =
            Operation.Update(dev.typr.foundations.Fragment.upsertOne(table, codec.underlying, row, *conflictColumns))

        @JvmStatic
        fun <Row : Any> upsertMany(
            table: String, codec: RowCodecNamed<Row>, rows: Iterator<Row>, vararg conflictColumns: String
        ): Operation.BatchUpdate<Row> =
            Operation.BatchUpdate(dev.typr.foundations.Fragment.upsertMany(table, codec.underlying, rows, *conflictColumns))

        @JvmStatic
        fun <Row : Any> upsertOneReturning(
            table: String, codec: RowCodecNamed<Row>, row: Row, vararg conflictColumns: String
        ): OperationRead.Query<Row> =
            OperationRead.Query(dev.typr.foundations.Fragment.upsertOneReturning(table, codec.underlying, row, *conflictColumns))

        @JvmStatic
        fun <Row : Any> insertIgnoreOne(
            table: String, codec: RowCodecNamed<Row>, row: Row, vararg conflictColumns: String
        ): Operation.Update =
            Operation.Update(dev.typr.foundations.Fragment.insertIgnoreOne(table, codec.underlying, row, *conflictColumns))

        @JvmStatic
        fun <Row : Any> insertIgnoreMany(
            table: String, codec: RowCodecNamed<Row>, rows: Iterator<Row>, vararg conflictColumns: String
        ): Operation.BatchUpdate<Row> =
            Operation.BatchUpdate(dev.typr.foundations.Fragment.insertIgnoreMany(table, codec.underlying, rows, *conflictColumns))

        @JvmStatic
        @JvmName("rowStatic")
        fun <Row : Any> row(codec: RowCodecNamed<Row>, row: Row, vararg except: String): Fragment =
            Fragment(dev.typr.foundations.Fragment.EMPTY.row(codec.underlying, row, *except))
    }
}

internal object OptionallyTransforms {
    val nullableToOptional: (Any?) -> Any? = { java.util.Optional.ofNullable(it) }

    val pairToOptionalTuple2: (Any?) -> Any? = { value ->
        if (value is Pair<*, *>) {
            java.util.Optional.of(dev.typr.foundations.Tuple.of(value.first, value.second))
        } else {
            java.util.Optional.empty<Any>()
        }
    }

    val tripleToOptionalTuple3: (Any?) -> Any? = { value ->
        if (value is Triple<*, *, *>) {
            java.util.Optional.of(dev.typr.foundations.Tuple.of(value.first, value.second, value.third))
        } else {
            java.util.Optional.empty<Any>()
        }
    }
}

File: foundations-jdbc-scala/src/scala/dev/typr/foundationssc/Fragment.scala
package dev.typr.foundationssc

import java.sql.PreparedStatement
import java.util.concurrent.atomic.AtomicInteger
import _root_.scala.jdk.CollectionConverters.*

/** Scala wrapper for dev.typr.foundations.Fragment with Scala-native APIs.
  *
  * This class wraps the Java Fragment interface and provides Scala-friendly methods that use Scala collections and types.
  */
class Fragment(val underlying: dev.typr.foundations.Fragment) extends AnyVal {

  def render(): String = underlying.render()

  def render(sb: java.lang.StringBuilder): Unit = underlying.render(sb)

  def set(stmt: PreparedStatement): Unit = underlying.set(stmt)

  def set(stmt: PreparedStatement, idx: AtomicInteger): Unit = underlying.set(stmt, idx)

  def append(other: Fragment): Fragment = new Fragment(underlying.append(other.underlying))

  def ++(other: Fragment): Fragment = append(other)

  def query[T](parser: ResultSetParser[T]): OperationRead.Query[T] =
    OperationRead.Query(this, parser)

  def queryExactlyOne[T](tpe: DbType[T]): OperationRead.Query[T] =
    query(RowCodec.of(tpe).exactlyOne())

  def queryExactlyOne[T](codec: RowCodec[T]): OperationRead.Query[T] =
    query(codec.exactlyOne())

  def queryAll[T](tpe: DbType[T]): OperationRead.Query[List[T]] =
    query(RowCodec.of(tpe).all())

  def queryAll[T](codec: RowCodec[T]): OperationRead.Query[List[T]] =
    query(codec.all())

  def queryMaxOne[T](tpe: DbType[T]): OperationRead.Query[Option[T]] =
    query(RowCodec.of(tpe).maxOne())

  def queryMaxOne[T](codec: RowCodec[T]): OperationRead.Query[Option[T]] =
    query(codec.maxOne())

  def update(): Operation.Update =
    new Operation.Update(new dev.typr.foundations.Operation.Update(underlying))

  def execute(): Operation.Execute =
    new Operation.Execute(new dev.typr.foundations.Operation.Execute(underlying))

  def updateReturning[T](parser: ResultSetParser[T]): Operation.UpdateReturning[T] =
    new Operation.UpdateReturning(new dev.typr.foundations.Operation.UpdateReturning(underlying, parser.underlying))

  def updateReturningGeneratedKeys[T](columnNames: Array[String], parser: ResultSetParser[T]): Operation.UpdateReturningGeneratedKeys[T] =
    new Operation.UpdateReturningGeneratedKeys(underlying.updateReturningGeneratedKeys(columnNames, parser.underlying))

  def updateMany[Row](parser: RowCodec[Row], rows: Iterator[Row]): Operation.UpdateMany[Row] = {
    import _root_.scala.jdk.CollectionConverters.*
    new Operation.UpdateMany(underlying.updateMany(parser.underlying, rows.asJava))
  }

  def updateManyReturning[Row](parser: RowCodec[Row], rows: Iterator[Row]): Operation.UpdateManyReturning[Row] = {
    import _root_.scala.jdk.CollectionConverters.*
    new Operation.UpdateManyReturning(underlying.updateManyReturning(parser.underlying, rows.asJava))
  }

  def updateReturningEach[Row](parser: RowCodec[Row], rows: Iterator[Row]): Operation.UpdateReturningEach[Row] = {
    import _root_.scala.jdk.CollectionConverters.*
    new Operation.UpdateReturningEach(underlying.updateReturningEach(parser.underlying, rows.asJava))
  }

  def streamingQuery[T](codec: RowCodec[T], fetchSize: Int): OperationRead.Streaming[T] =
    new OperationRead.Streaming(underlying.streamingQuery(codec.underlying, fetchSize))

  def streamingQuery[T](tpe: DbType[T], fetchSize: Int): OperationRead.Streaming[T] =
    new OperationRead.Streaming(underlying.streamingQuery(tpe.underlying, fetchSize))

  def append(s: String): Fragment = new Fragment(underlying.append(s))

  def value[T](dbType: DbType[T], value: T): Fragment = new Fragment(underlying.value(dbType.underlying, value))

  def appendAll(fragments: List[Fragment], separator: Fragment): Fragment =
    new Fragment(underlying.appendAll(fragments.map(_.underlying).asJava, separator.underlying))

  def valueNullable[T](dbType: DbType[T], value: Option[T]): Fragment = {
    import _root_.scala.jdk.OptionConverters.*
    new Fragment(underlying.value(dbType.underlying.opt(), value.toJava))
  }

  def row[Row](parser: RowCodecNamed[Row], row: Row, except: String*): Fragment =
    new Fragment(underlying.row(parser.underlying, row, except*))

  def param[P0](dbType: DbType[P0]): ParamBuilders.ParamBuilder1[P0] =
    new ParamBuilders.ParamBuilder1(underlying.param(dbType.underlying))

  // ── Conditional append DSL ──

  def optionally[T](value: Option[T]): Fragment.OptionallyValue[T] =
    new Fragment.OptionallyValue(this, value)

  def optionally(condition: Boolean): Fragment.OptionallyFlag =
    new Fragment.OptionallyFlag(this, condition)
}

object Fragment {
  class OptionallyValue[T](base: Fragment, value: Option[T]):
    def append(sql: String, tpe: DbType[T]): Fragment =
      import _root_.scala.jdk.OptionConverters.*
      new Fragment(
        base.underlying
          .optionally(value.toJava.asInstanceOf[java.util.Optional[T]])
          .append(sql, tpe.underlying))

    def append(sql: String, tpe: DbType[T], whenAbsent: String): Fragment =
      import _root_.scala.jdk.OptionConverters.*
      new Fragment(
        base.underlying
          .optionally(value.toJava.asInstanceOf[java.util.Optional[T]])
          .append(sql, tpe.underlying, whenAbsent))

  class OptionallyFlag(base: Fragment, condition: Boolean):
    def append(sql: String): Fragment =
      new Fragment(base.underlying.optionally(condition).append(sql))

    def append(fragment: Fragment): Fragment =
      new Fragment(base.underlying.optionally(condition).append(fragment.underlying))

    def append(whenTrue: String, whenFalse: String): Fragment =
      new Fragment(base.underlying.optionally(condition).append(whenTrue, whenFalse))

    def append(whenTrue: Fragment, whenFalse: Fragment): Fragment =
      new Fragment(base.underlying.optionally(condition).append(whenTrue.underlying, whenFalse.underlying))

  val EMPTY: Fragment = new Fragment(dev.typr.foundations.Fragment.EMPTY)

  def of(value: String): Fragment = new Fragment(dev.typr.foundations.Fragment.of(value))

  def empty(): Fragment = EMPTY

  def quotedDouble(value: String): Fragment = new Fragment(dev.typr.foundations.Fragment.quotedDouble(value))

  def quotedSingle(value: String): Fragment = new Fragment(dev.typr.foundations.Fragment.quotedSingle(value))

  def value[A](value: A, dbType: DbType[A]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.value(value, dbType.underlying))

  def encode[A](dbType: DbType[A], value: A): Fragment =
    new Fragment(dev.typr.foundations.Fragment.encode(dbType.underlying, value))

  def and(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.and(fragments.map(_.underlying)*))

  def and(fragments: List[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.and(fragments.map(_.underlying).asJava))

  def or(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.or(fragments.map(_.underlying)*))

  def or(fragments: List[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.or(fragments.map(_.underlying).asJava))

  def whereAnd(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.whereAnd(fragments.map(_.underlying)*))

  def whereAnd(fragments: List[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.whereAnd(fragments.map(_.underlying).asJava))

  def whereOr(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.whereOr(fragments.map(_.underlying)*))

  def whereOr(fragments: List[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.whereOr(fragments.map(_.underlying).asJava))

  def set(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.set(fragments.map(_.underlying)*))

  def set(fragments: List[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.set(fragments.map(_.underlying).asJava))

  def parentheses(fragment: Fragment): Fragment =
    new Fragment(dev.typr.foundations.Fragment.parentheses(fragment.underlying))

  def comma(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.comma(fragments.map(_.underlying)*))

  def comma(fragments: Iterable[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.comma(fragments.map(_.underlying).toList.asJava))

  def orderBy(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.orderBy(fragments.map(_.underlying)*))

  def orderBy(fragments: List[Fragment]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.orderBy(fragments.map(_.underlying).asJava))

  def join(fragments: List[Fragment], separator: Fragment): Fragment =
    new Fragment(dev.typr.foundations.Fragment.join(fragments.map(_.underlying).asJava, separator.underlying))

  def concat(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.concat(fragments.map(_.underlying)*))

  /** Scala string interpolator for creating SQL Fragments. */
  extension (sc: StringContext) {
    def sql(args: Fragment*): Fragment = {
      val parts = sc.parts.iterator
      val frags = new scala.collection.mutable.ListBuffer[dev.typr.foundations.Fragment]()

      if (parts.hasNext) {
        val first = parts.next()
        if (first.nonEmpty) {
          frags += dev.typr.foundations.Fragment.of(first)
        }
      }

      val argsIt = args.iterator
      while (parts.hasNext && argsIt.hasNext) {
        frags += argsIt.next().underlying
        val part = parts.next()
        if (part.nonEmpty) {
          frags += dev.typr.foundations.Fragment.of(part)
        }
      }

      while (argsIt.hasNext) {
        frags += argsIt.next().underlying
      }

      frags.result() match {
        case Nil           => Fragment.empty()
        case single :: Nil => new Fragment(single)
        case multiple      =>
          val javaList = new java.util.ArrayList[dev.typr.foundations.Fragment](multiple.size)
          multiple.foreach(javaList.add)
          new Fragment(new dev.typr.foundations.Fragment.Concat(javaList))
      }
    }
  }

  def of(fragments: Fragment*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.of(fragments.map(_.underlying)*))

  /** Emit `DROP TABLE IF EXISTS <table>`. Works on PostgreSQL, DuckDB, MariaDB, MySQL, SQL Server (2016+) and Oracle (23c+). DB2 does not support the
    * `IF EXISTS` clause — wrap the plain `DROP TABLE <table>` in a try/catch for SQLSTATE 42704.
    */
  def dropTableIfExists(table: String): Operation.Execute =
    new Operation.Execute(dev.typr.foundations.Fragment.dropTableIfExists(table))

  def insertOne[Row](table: String, codec: RowCodecNamed[Row], row: Row, except: String*): Operation.Update =
    new Operation.Update(dev.typr.foundations.Fragment.insertOne(table, codec.underlying, row, except*))

  def insertMany[Row](table: String, codec: RowCodecNamed[Row], rows: Iterator[Row], except: String*): Operation.BatchUpdate[Row] =
    new Operation.BatchUpdate(dev.typr.foundations.Fragment.insertMany(table, codec.underlying, rows.asJava, except*))

  def insertOneReturning[Row](table: String, codec: RowCodecNamed[Row], row: Row, except: String*): OperationRead.Query[Row] =
    new OperationRead.Query(dev.typr.foundations.Fragment.insertOneReturning(table, codec.underlying, row, except*))

  def insertOneReturning[In, Out](table: String, writeCodec: RowCodecNamed[In], row: In, readCodec: RowCodecNamed[Out]): OperationRead.Query[Out] =
    new OperationRead.Query(dev.typr.foundations.Fragment.insertOneReturning(
      table, writeCodec.underlying, row, readCodec.underlying))

  def insertOneGenerated[Row, Out](
      table: String, codec: RowCodecNamed[Row], row: Row,
      generatedColumns: Array[String], parser: ResultSetParser[Out], except: String*
  ): Operation.UpdateReturningGeneratedKeys[Out] =
    new Operation.UpdateReturningGeneratedKeys(dev.typr.foundations.Fragment.insertOneGenerated(
      table, codec.underlying, row, generatedColumns, parser.underlying, except*))

  def upsertOne[Row](table: String, codec: RowCodecNamed[Row], row: Row, conflictColumns: String*): Operation.Update =
    new Operation.Update(dev.typr.foundations.Fragment.upsertOne(table, codec.underlying, row, conflictColumns*))

  def upsertMany[Row](table: String, codec: RowCodecNamed[Row], rows: Iterator[Row], conflictColumns: String*): Operation.BatchUpdate[Row] =
    new Operation.BatchUpdate(dev.typr.foundations.Fragment.upsertMany(table, codec.underlying, rows.asJava, conflictColumns*))

  def upsertOneReturning[Row](table: String, codec: RowCodecNamed[Row], row: Row, conflictColumns: String*): OperationRead.Query[Row] =
    new OperationRead.Query(dev.typr.foundations.Fragment.upsertOneReturning(table, codec.underlying, row, conflictColumns*))

  def insertIgnoreOne[Row](table: String, codec: RowCodecNamed[Row], row: Row, conflictColumns: String*): Operation.Update =
    new Operation.Update(dev.typr.foundations.Fragment.insertIgnoreOne(table, codec.underlying, row, conflictColumns*))

  def insertIgnoreMany[Row](table: String, codec: RowCodecNamed[Row], rows: Iterator[Row], conflictColumns: String*): Operation.BatchUpdate[Row] =
    new Operation.BatchUpdate(dev.typr.foundations.Fragment.insertIgnoreMany(table, codec.underlying, rows.asJava, conflictColumns*))

  def row[Row](codec: RowCodecNamed[Row], row: Row, except: String*): Fragment =
    new Fragment(dev.typr.foundations.Fragment.EMPTY.row(codec.underlying, row, except*))

  def valuesList[A](dbType: DbType[A], values: Iterable[A]): Fragment =
    new Fragment(dev.typr.foundations.Fragment.valuesList(dbType.underlying, values.asJava))
}
